synapse/events/snapshot.py

```diff
@@ -482,6 +461,34 @@ def _encode_state_dict(
     return [(etype, state_key, v) for (etype, state_key), v in state_dict.items()]
 
 
+def _encode_state_group_delta(
+    state_group_delta: Optional[Dict[Tuple[int, int], StateMap[str]]]
+) -> Optional[List[Tuple[int, int, Optional[List[Tuple[str, str, str]]]]]]:
+    if not state_group_delta:
+        return None
+
+    state_group_delta_encoded = []
+    for key, value in state_group_delta.items():
+        state_group_delta_encoded.append((key[0], key[1], _encode_state_dict(value)))
+
+    return state_group_delta_encoded
+
+
+def _decode_state_group_delta(
+    input: Optional[List[Tuple[int, int, List[Tuple[str, str, str]]]]]
+) -> Optional[Dict[Tuple[int, int], StateMap[str]]]:
+    if not input:
+        return None
+
+    state_group_deltas = {}
+    for element in input:
+        state_map = _decode_state_dict(element[2])
+        assert state_map is not None
+        state_group_deltas[(element[0], element[1])] = state_map
+
+    return state_group_deltas
```

Comment:
Please move these above `_encode_state_dict` (so that the pairs of encode/decode functions are together).